src/query/parser/promql/types.go

```diff
@@ -424,3 +424,29 @@ func promMatchingToM3(
 		Include:        vectorMatching.Include,
 	}
 }
+
+// Iff one of left or right is a time() function, match match one to many
+// against it, and match everything.
+func timeMatcher(left, right bool) *binary.VectorMatching {
```

Comment:
Similar, maybe here we make it return `(*binary.VectorMatching, bool)`?

Even if we don't use the output, still makes the top level look better:
```go
matching, ok := promMatchingToM3(expr.VectorMatching)
if !ok {
  matching, _ = timeMatcher(isTime(expr.LHS), isTime(expr.RHS))
}

```